frame/composable-traits/src/staking/mod.rs

```diff
@@ -1,35 +1,138 @@
+use crate::time::DurationSeconds;
+use frame_support::{dispatch::DispatchResult, pallet_prelude::*, BoundedBTreeMap};
 use scale_info::TypeInfo;
+use sp_arithmetic::traits::Zero;
 use sp_runtime::{DispatchError, Perbill, Permill};
 
 pub mod lock;
 pub mod math;
 pub mod nft;
-pub mod rewards;
 
-#[derive(Debug, PartialEq, Eq, Clone, Encode, Decode, TypeInfo)]
-pub struct StakeConfig<DurationPresets, RewardsRate> {
-	pub lock: Option<LockConfig<DurationPresets>>,
-	pub reward: Option<RewardConfig<RewardsRate>>,
+pub type StakingDurationToRewardsMultiplierConfig<Limit> =
+	BoundedBTreeMap<DurationSeconds, Perbill, Limit>;
+
+/// Defines staking duration, rewards and early unstake penalty for a given asset type.
+/// TODO refer to the relevant section in the design doc.
+#[derive(RuntimeDebug, PartialEq, Eq, Clone, Encode, Decode, TypeInfo)]
+pub struct Reward<AssetId, Balance> {
+	/// asset id of the reward
+	pub asset_id: AssetId,
+
+	/// Total rewards including inflation for adjusting for new stakers joining the pool. All
+	/// stakers in a pool are eligible to receive a part of this value based on their share of the
+	/// pool.
+	pub total_rewards: Balance,
+
+	/// A book keeping field to track the actual total reward without the
+	/// reward dilution adjustment caused by new stakers joining the pool.
+	pub total_dilution_adjustment: Balance,
+
+	/// Upper bound on the `total_rewards - total_dilution_adjustment`.
+	pub max_rewards: Balance,
+
+	/// The rewarding rate that increases the pool `total_reward` (and `actual_total_reward`)
+	/// at a given time.
+	pub reward_rate: Perbill,
 }
 
-/// staking typed fNFT, usually can be mapped to raw fNFT storage type
-#[derive(Debug, PartialEq, Eq, Copy, Clone, Encode, Decode, TypeInfo)]
-pub struct Stake<Balance, Rewards> {
+impl<AssetId, Balance: Zero> Reward<AssetId, Balance> {
+	pub fn from(reward_config: RewardConfig<AssetId, Balance>) -> Reward<AssetId, Balance> {
+		Reward {
+			asset_id: reward_config.asset_id,
+			total_rewards: Zero::zero(),
+			total_dilution_adjustment: Zero::zero(),
+			max_rewards: reward_config.max_rewards,
+			reward_rate: reward_config.reward_rate,
+		}
+	}
+}
+
+/// A reward pool is a collection of rewards that are allocated to stakers to incentivize a
+/// particular purpose. Eg: a pool of rewards for incentivizing adding liquidity to a pablo swap
+/// pool. TODO refer to the relevant section in the design doc.
+#[derive(RuntimeDebug, PartialEq, Eq, Clone, Encode, Decode, TypeInfo)]
+pub struct RewardPool<
+	AccountId,
+	AssetId,
+	Balance,
+	BlockNumber,
+	DurationPresets,
+	RewardsLength: Get<u32>,
+> {
+	pub owner: AccountId,
+
+	/// The staked asset id of the reward pool.
+	pub asset_id: AssetId,
+
+	/// rewards accumulated
+	pub rewards: BoundedBTreeMap<AssetId, Reward<AssetId, Balance>, RewardsLength>,
+
+	/// Total shares distributed among stakers
+	pub total_shares: Balance,
+
+	/// Pool would stop adding rewards to pool at this block number.
```

Comment:
we have rate,  stop block and max rewards. so if rewards max will be reached before block stop, than effectively max is less than expected. how would you handel that in UI to end users? i proposed solution which will not have such an issue

Reply:
soltion was to have only reward rate and block stop. not max

Reply:
That is totally fine those are independent variables. If there's a lot of staking than we expect then the max will be reached earlier

Reply:
how max will be reached earlier? what if max reached, and end_block not. so we have end_block not reached, but there is no rewards. it is noncece.

Reply:
It's not non-sense, there are two ways to control the rewards campaign, either it runs out of time or runs out of funds, perfectly normal.

Reply:
max rewads = time * rate. having to ways of control does not makes any sense except creating ambiguity. 
one may choose reward amount, then get rate to fit that.

Reply:
please provide concrete example when it will not be the case

Reply:
why to introduce stuff when there is no use case?

Reply:
Let me explain one last time on top of many other times I have explained to you online/offline/chat :sweat_smile: 

- max rewards = time * rate  + any additional transferred rewards
- `rate` is adjusted(governance/gauge etc for incentivizing staking or the opposite) so it's not possible to predict that end_block will end with the max reward possible
- `any additional transferred rewards` increases . Then the pool can satisfy it's maximum reward goal earlier than expected, then there is no need to keep. To control this scenario we need a max reward independent of the end block/time. 
- Also the rewarded assets are independent of each other, one can reach the max reward for it before the others and before the overall reward pool has expired.
- You can see this type of dynamic uniswap clones, most recent one I used is https://app.stellaswap.com/farm

Reply:
I think that design is fine. Governance just needs the knobs to actually configure this.

Reply:
4. about assets and pool have different lifecycles. i did not say differently. each asset can have its end block.

Reply:
2. not sure why need to set max to reward explicitly?  why would somebody adding reward into the pool after there is  no need to do so? will be other pallet?  
and i do propose to predict anything.

Reply:
3. not sure. in those case should not rate to be set to zero? so pool is not rewardinf anymore?

Reply:
i imagine UI. reward rate is nice, end block is far ahead. and pool not rewarding at all. but wait, there is reward rate, why pool not rewarding?

Reply:
it possible, that hook logic for rate could track previous reward and calculated by per asset block. if somebody rewarded our of rate via extrinsic, it can minus block for proper amount.   no need for max to ve stored on chain and  no incosistencies in ui. so max reward = already rewarded + rate * remaining time.

Reply:
so what i say that max reward is derived field and not having it in state will make UI more consistent.